engine/src/main/java/org/terasology/world/chunks/blockdata/ExtraBlockDataManager.java

```diff
@@ -0,0 +1,320 @@
+                        }
+                    }
+                    fieldss.get(registerAnnotation.bitSize()).put(registerAnnotation.name(), includedBlocks);
+                }
+            }
+        }
+        
+        slots = new HashMap<>();
+        ArrayList<TeraArray.Factory<?>> tempSlotTypes = new ArrayList<>();
+        fieldss.forEach((size, fields) -> {
+            Graph disjointnessGraph = getDisjointnessGraph(fields);
+            ArrayList<ArrayList<String>> cliques = findCliqueCover(disjointnessGraph);
+            for (ArrayList<String> clique : cliques) {
+                for (String label : clique) {
+                    slots.put(label, tempSlotTypes.size());
+                }
+                tempSlotTypes.add(TERA_ARRAY_FACTORIES.get(size));
+            }
+        });
+        slotFactories = tempSlotTypes.toArray(new TeraArray.Factory<?>[0]);
+        String loggingOutput = "Extra data slots registered:";
+        boolean first = true;
+        for (Map.Entry<String, Integer> entry : slots.entrySet()) {
+            loggingOutput += (first ? " " : ", ") + entry.getKey() + " -> " + entry.getValue();
+            first = false;
+        }
+        logger.info(loggingOutput);
+    }
+    
+    private static String validRegistrationMethod(Method method, RegisterExtraData annotation) {
+        Class<?>[] argumentTypes = method.getParameterTypes();
+        return method.getReturnType() != boolean.class ? "incorrect return type"
+               : !TERA_ARRAY_FACTORIES.containsKey(annotation.bitSize()) ? "invalid bitSize"
+               : !Modifier.isStatic(method.getModifiers()) ? "method not static"
+               : argumentTypes.length != 1 ? "arguments list has wrong length"
+               : argumentTypes[0] != Block.class ? "incorrect argument type"
+               : null;
+    }
+    
+    private static Graph getDisjointnessGraph(Map<String, Set<Block>> fields) {
+        Graph graph = new Graph(fields.keySet().toArray(new String[0]));
+        fields.forEach((name0, blockSet0) ->
+            fields.forEach((name1, blockSet1) -> {
+                if (name0.compareTo(name1) < 0 && isDisjoint(blockSet0, blockSet1)) {
+                    graph.addEdge(name0, name1);
+                }
+            })
+        );
+        return graph;
+    }
+    
+    private static <T> boolean isDisjoint(Set<T> s0, Set<T> s1) {
+        for (T t : s0) {
+            if (s1.contains(t)) {
+                return false;
+            }
+        }
+        return true;
+    }
+    
+    //This is exponential time, but the problem is known to be NP-hard in general and large cases are unlikely to come up.
+    /**
+     * Find the smallest-possible number of cliques that contain all of the vertices of the graph.
+     * Used to determine how to most efficiently assign aliases of requested extra-data fields.
+     */
+    private static ArrayList<ArrayList<String>> findCliqueCover(Graph graph) {
```

Comment:
As in, just split the constructor into parts? It would seem to serve that purpose just as well to add a few explanatory comments to the constructor (e.g. "`// Find requests for extensions and which blocks they apply to`" and "`// Work out which fields don't overlap and can be aliased together.`" at the starts of those sections).

Reply:
The constructor is quite big and complex. It seems to consists of several (non-trivial) sections.
It would be better if these sections are more clearly labeled.

Using comments to label the sections is also OK.
However, if you use a named function, then the function signature probably tells me  (and the compiler and IDE!)  a lot more then multiple lines of comments could.

Reply:
That seems reasonable. The function signature for the aliasing part would be less helpful, as that sets up multiple fields (as in Java fields, not extra-data fields), so at least part of its operation would end up as side-effects.